feat: notifying  incosistent fs version problem with exit code

PerseidMeteor opened the pull request:
If acceld converts with different fs version cache, leading to an inconsistent fs version problem when merging into boostrap layer. So we need to notify acceld that an inconsistent version occured and handle this error.

## Relevant Issue (if applicable)
_If there are Issues related to this PullRequest, please list it._

## Details
Define `InconsistentFsVersion` with `thiserror` and check error type after merge, if is fs version inconsistent problem, just return status code 2.

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

imeoer commented on rafs/src/metadata/mod.rs:
```diff
@@ -397,34 +399,44 @@ pub struct RafsSuperConfig {
 }
 
+#[derive(Error, Debug)]
+pub enum MergeError {
```
How about just using this:

```
pub enum MergeError {
    #[error("Inconsistent RAFS Filesystem: {0}")]
    InconsisentFilesystem(String),
    #[error(transparent)]
    Other(#[from] Error),
}
```

PerseidMeteor replied:
ok, I droped the `DifferentDigestAlgorithm(String)`

imeoer commented on builder/src/merge.rs:
```diff
@@ -83,6 +83,7 @@ impl Merger {
         target: ArtifactStorage,
         chunk_dict: Option<PathBuf>,
         config_v2: Arc<ConfigV2>,
+        expected_fs_version: Option<u32>,
```
It seems we need to also consider other options like `chunk_size`, `digester`, I think maybe it's unnecessary to check this.

PerseidMeteor replied:
Dropped this commit temporarily

PerseidMeteor commented on src/bin/nydus-image/main.rs:
```diff
@@ -759,7 +759,13 @@ fn main() -> Result<()> {
     } else if let Some(matches) = cmd.subcommand_matches("merge") {
-        Command::merge(matches, &build_info)
+        let result = Command::merge(matches, &build_info);
+        if let Err(e) = result {
```
we may can do it `if let MergeError::InconsisentFsVersion(_) = result.unwrap_err()`

imeoer commented on rafs/src/metadata/mod.rs:
```diff
@@ -397,34 +398,42 @@ pub struct RafsSuperConfig {
+#[derive(Error, Debug)]
+pub enum MergeError {
+    #[error("Inconsistent RAFS version:{0}")]
+    InconsisentFsVersion(String),
```
Better be

```
#[error("Inconsistent RAFS Filesystem: {0}")]
nconsisentFilesystem(String),
```

PerseidMeteor replied:
There will be a warning if doesn't have an upper camel case name.May be follows better.
```
#[error("Inconsistent RAFS Filesystem: {0}")]
NconsisentFilesystem(String),
```

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -759,7 +759,11 @@ fn main() -> Result<()> {
+        if let Err(MergeError::InconsisentFsVersion(_)) = result {
+            std::process::exit(2);
+        }
+        Ok(())
```
Other errors should also be handled here?

PerseidMeteor replied:
yep

imeoer commented on rafs/src/metadata/mod.rs:
```diff
@@ -397,34 +398,44 @@ pub struct RafsSuperConfig {
+#[derive(Error, Debug)]
+pub enum MergeError {
+    #[error("Inconsistent RAFS Filesystem: {0}")]
+    NconsisentFilesystem(String),
```
`InconsistentFilesystem(String),` ?

imeoer commented on rafs/src/metadata/mod.rs:
```diff
@@ -397,34 +398,44 @@ pub struct RafsSuperConfig {
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+    #[error(transparent)]
+    IOError(#[from] std::io::Error),
```
Is it necessary?

PerseidMeteor replied:
yep, because these two func return an `std::io::Error`. If we don't have an `IOError`  from, they can not be converted to an MergeError automaticly.

imeoer replied:
But it seems also to be converted to `MergeError::Other` in `OutputSerializer::dump(matches, output, build_info).map_err(MergeError::Other)`.

PerseidMeteor replied:
> yep, because these two func return an `std::io::Error`. If we don't have an `IOError` from, they can not be converted to an MergeError automaticly.

The func is `blob_ctx.blob_id =
                                BlobInfo::get_blob_id_from_meta_path(bootstrap_path)?;`
We can manually convert it to `MergeError.`

imeoer commented on builder/src/merge.rs:
```diff
@@ -83,49 +83,56 @@ impl Merger {
-                sources.len(),
-            );
+            if digests.len() != sources.len() {
+                Err(MergeError::Other(anyhow!(
```
Does this let the `merge` func return the error immediately?

PerseidMeteor replied:
maybe we can do it use `return`

PerseidMeteor replied:
Maybe we can use `return`.

imeoer commented on builder/src/merge.rs:
```diff
@@ -222,8 +234,11 @@ impl Merger {
-                            blob_ctx.blob_id =
-                                BlobInfo::get_blob_id_from_meta_path(bootstrap_path)?;
+                            let blob_id = BlobInfo::get_blob_id_from_meta_path(bootstrap_path);
+                            if let Err(err) = blob_id {
```
Better to use `map_err` ?

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -758,7 +758,11 @@ fn main() -> Result<()> {
-        Command::merge(matches, &build_info)
+        let result = Command::merge(matches, &build_info);
+        if let Err(MergeError::InconsistentFilesystem(_)) = result {
+            std::process::exit(2);
```
Does this will print the error message to stderr before exit?

PerseidMeteor replied:
it seems not,may be use `eprintln`?

PerseidMeteor committed:
feat: notifying incosistent fs version problem with exit code
If acceld converts with different fs version cache, leading to an inconsistent fs version problem when merging into boostrap layer. So we need to notify acceld that an inconsistent version occured and handle this error.

Signed-off-by: YuQiang <y_q_email@163.com>

imeoer reviewed (approved):
LGTM! Thanks!
